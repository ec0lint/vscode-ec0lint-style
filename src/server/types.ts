import type { Connection } from 'vscode-languageserver';
import type { TextDocuments } from 'vscode-languageserver/node';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
// import { CodeActionKind as VSCodeActionKind } from 'vscode-languageserver-types';
import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type winston from 'winston';
import type { Ec0lintRunner as Ec0lintStyleRunner, LintDiagnostics } from '../utils/ec0lint-style';
import type { ExtractKeysOfValueType } from '../utils/types';
import type { PackageManager, Ec0lintStyleResolutionResult } from '../utils/packages';
import type { NotificationManager, CommandManager } from '../utils/lsp';

/**
 * Command IDs
 */
export enum CommandId {
	OpenRuleDoc = 'ec0lint-style.openRuleDoc',
}

/**
 * Language server notification names.
 */
export enum Notification {
	DidRegisterCodeActionRequestHandler = 'ec0lint-style/didRegisterCodeActionRequestHandler',
	DidRegisterDocumentFormattingEditProvider = 'textDocument/didRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'ec0lint-style/didResetConfiguration',
}

/**
 * `DidRegisterDocumentFormattingEditProvider` notification parameters.
 */
export interface DidRegisterDocumentFormattingEditProviderNotificationParams {
	/**
	 * The URI of the document for which the formatting provider was registered.
	 */
	readonly uri: string;

	/**
	 * The options used to register the document formatting provider.
	 */
	readonly options: LSP.DocumentFormattingRegistrationOptions;
}

/**
 * Context shared between the language server and its modules.
 */
export interface LanguageServerContext {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The notification manager for the connection.
	 */
	notifications: NotificationManager;

	/**
	 * The command manager for the connection.
	 */
	commands: CommandManager;

	/**
	 * The text document manager.
	 */
	documents: TextDocuments<TextDocument>;

	/**
	 * The runner with which to run ec0lint-style.
	 */
	runner: Ec0lintStyleRunner;

	/**
	 * Displays the given error in the client using the language server
	 * connection.
	 * @param error The error to display.
	 */
	displayError(error: unknown): void;

	/**
	 * Gets the effective extension options for a resource, given its URI.
	 * @param uri The resource URI.
	 */
	getOptions(uri: string): Promise<LanguageServerOptions>;

	/**
	 * Returns the module with the given ID if it exists.
	 * @param id The ID of the module to return.
	 */
	getModule(id: string): LanguageServerModule | undefined;

	/**
	 * Lints a document using ec0lint-style and returns fix text edits.
	 * @param document The document to get text edits for.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	getFixes(document: TextDocument, linterOptions?: Ec0lintStyle.LinterOptions): Promise<TextEdit[]>;

	/**
	 * Lints a document using ec0lint-style and returns diagnostics.
	 * @param document The document to lint.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	lintDocument(
		document: TextDocument,
		linterOptions?: Partial<Ec0lintStyle.LinterOptions>,
	): Promise<LintDiagnostics | undefined>;

	/**
	 * Resolves the ec0lint-style package to be used for the given document.
	 * @param document The document to resolve the package for.
	 */
	resolveEc0lintStyle(document: TextDocument): Promise<Ec0lintStyleResolutionResult | undefined>;
}

/**
 * A language server module.
 */
export interface LanguageServerModule {
	/**
	 * Handler called when the server is initializing.
	 */
	onInitialize?: (params: LSP.InitializeParams) => Partial<LSP.InitializeResult> | undefined | void;

	/**
	 * Handler called after the language server finishes registering handlers
	 * with the connection.
	 */
	onDidRegisterHandlers?: () => void;

	/**
	 * Handler called after the language server has finished responding to the
	 * onDidChangeConfiguration event.
	 */
	onDidChangeConfiguration?: () => void;

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[key: string | symbol]: any;
}

/**
 * A language server module class.
 */
export interface LanguageServerModuleConstructor {
	/**
	 * The module's ID, used to identify the module in the language server's
	 * internal state and when logging. Should be a unique, short, lowercase
	 * string.
	 */
	id: string;
	new (params: LanguageServerModuleConstructorParameters): LanguageServerModule;
}

/**
 * Parameters for the {@link LanguageServerModuleConstructor} constructor.
 */
export type LanguageServerModuleConstructorParameters = {
	context: LanguageServerContext;
	logger?: winston.Logger;
};

/**
 * Language server event handler names.
 */
export type LanguageServerHandlers = ExtractKeysOfValueType<
	LanguageServerModule,
	() => unknown | undefined
>;

/**
 * Parameters for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerParameters = {
	[key in LanguageServerHandlers]: Parameters<Required<LanguageServerModule>[key]>;
};

/**
 * Return types for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerReturnValues = {
	[key in LanguageServerHandlers]: ReturnType<Required<LanguageServerModule>[key]>;
};

/**
 * Language server constructor parameters.
 */
export type LanguageServerConstructorParameters = {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The logger to use.
	 */
	logger?: winston.Logger;

	/**
	 * The modules to load.
	 */
	modules?: LanguageServerModuleConstructor[];
};

/**
 * Language server options.
 */
export type LanguageServerOptions = {
	codeAction: {
		disableRuleComment: {
			location: 'separateLine' | 'sameLine';
		};
	};
	config?: Ec0lintStyle.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet: string[];
	ec0lintStylePath?: string;
	validate: string[];
};
