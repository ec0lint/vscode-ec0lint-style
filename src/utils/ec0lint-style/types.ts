import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { PackageManager } from '../packages';

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: Ec0lintStyle.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
