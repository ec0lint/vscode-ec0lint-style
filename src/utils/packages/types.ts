// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: Ec0lintStyle.PublicApi;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
