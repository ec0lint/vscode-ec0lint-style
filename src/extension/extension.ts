import { EventEmitter } from 'events';
import { LanguageClient, SettingMonitor, ExecuteCommandRequest } from 'vscode-languageclient/node';
import { workspace, commands, window } from 'vscode';
import { ApiEvent, PublicApi } from './types';
import {
	CommandId,
	DidRegisterDocumentFormattingEditProviderNotificationParams,
	Notification,
} from '../server';
import type vscode from 'vscode';

/**
 * Activates the extension.
 */
export function activate({ subscriptions }: vscode.ExtensionContext): PublicApi {
	const serverPath = require.resolve('./start-server');

	const api: PublicApi = Object.assign(new EventEmitter(), { codeActionReady: false });

	const client = new LanguageClient(
		'Ec0lintStyle',
		{
			run: {
				module: serverPath,
			},
			debug: {
				module: serverPath,
				options: {
					execArgv: ['--nolazy', '--inspect=6004'],
				},
			},
		},
		{
			documentSelector: [{ scheme: 'file' }, { scheme: 'untitled' }],
			diagnosticCollectionName: 'ec0lint-style',
			synchronize: {
				fileEvents: workspace.createFileSystemWatcher(
					'**/{.ec0lint-stylerc{,.js,.json,.yaml,.yml},ec0lint-style.config.js,.ec0lint-styleignore}',
				),
			},
		},
	);

	client
		.onReady()
		.then(() => {
			client.onNotification(Notification.DidRegisterCodeActionRequestHandler, () => {
				api.codeActionReady = true;
			});
			client.onNotification(
				Notification.DidRegisterDocumentFormattingEditProvider,
				(params: DidRegisterDocumentFormattingEditProviderNotificationParams) => {
					api.emit(ApiEvent.DidRegisterDocumentFormattingEditProvider, params);
				},
			);
			client.onNotification(Notification.DidResetConfiguration, () => {
				api.emit(ApiEvent.DidResetConfiguration);
			});
		})
		.catch(async (error) => {
			await window.showErrorMessage(
				`ec0lint-style: ${error instanceof Error ? error.message : String(error)}`,
			);
		});

	subscriptions.push(
		// cspell:disable-next-line
		commands.registerCommand('ec0lint-style.executeAutofix', async () => {
			const textEditor = window.activeTextEditor;

			if (!textEditor) {
				return;
			}

			const textDocument = {
				uri: textEditor.document.uri.toString(),
				version: textEditor.document.version,
			};
			const params = {
				command: CommandId.ApplyAutoFix,
				arguments: [textDocument],
			};

			await client.sendRequest(ExecuteCommandRequest.type, params).then(undefined, async () => {
				await window.showErrorMessage(
					'Failed to apply ec0lint-style fixes to the document. Please consider opening an issue with steps to reproduce.',
				);
			});
		}),
	);

	subscriptions.push(new SettingMonitor(client, 'ec0lint-style.enable').start());

	return api;
}
