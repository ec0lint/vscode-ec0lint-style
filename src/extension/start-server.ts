import path from 'path';
import { createConnection, ProposedFeatures } from 'vscode-languageserver/node';
import winston from 'winston';
import { Ec0lintStyleLanguageServer as Ec0lintLanguageServer, modules } from '../server';

const connection = createConnection(ProposedFeatures.all);

import { ErrorFormatter, LanguageServerTransport, LanguageServerFormatter } from '../utils/logging';

const { NODE_ENV } = process.env;

const level = NODE_ENV === 'development' ? 'debug' : 'info';

const transports: winston.transport[] = [
	new LanguageServerTransport({
		connection,
		format: winston.format.combine(
			new ErrorFormatter(),
			new LanguageServerFormatter({
				connection,
				preferredKeyOrder: ['module', 'uri', 'command'],
			}),
		),
	}),
];

if (level === 'debug') {
	transports.push(
		new winston.transports.File({
			filename: path.join(__dirname, '../ec0lint-style-language-server.log'),
			level,
			format: winston.format.combine(
				new ErrorFormatter(),
				winston.format.timestamp(),
				winston.format.json(),
			),
		}),
	);
}

const logger = winston.createLogger({ level, transports });

const server = new Ec0lintLanguageServer({
	connection,
	logger,
	modules: Object.values(modules),
});

server.start();
