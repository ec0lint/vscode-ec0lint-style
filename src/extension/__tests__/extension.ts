jest.doMock('vscode', () => ({ workspace: {}, window: {}, commands: {} }), { virtual: true });
jest.mock('vscode-languageclient/node', () => ({
	LanguageClient: jest.fn(),
	SettingMonitor: jest.fn(),
	ExecuteCommandRequest: {
		type: 'executeCommand',
	},
}));

import { EventEmitter } from 'events';
import vscode, { window } from 'vscode';
import { LanguageClient, SettingMonitor, NodeModule } from 'vscode-languageclient/node';
import {
	DidRegisterDocumentFormattingEditProviderNotificationParams,
	Notification,
} from '../../server';
import { activate } from '../extension';
import { ApiEvent } from '../types';

const mockVSCode = vscode as jest.Mocked<typeof vscode>;
const mockLanguageClient = LanguageClient as jest.MockedClass<typeof LanguageClient>;
const mockSettingMonitor = SettingMonitor as jest.MockedClass<typeof SettingMonitor>;

const mockWorkspace = {
	createFileSystemWatcher: jest.fn(),
};

const mockCommands = {
	registerCommand: jest.fn(),
};

const mockTextEditor = {
	document: {
		uri: {
			toString: () => 'file:///path/to/file.ts',
		},
		version: 1,
	},
} as vscode.TextEditor;

const mockWindow = {
	activeTextEditor: mockTextEditor,
	showErrorMessage: jest.fn(),
};

mockVSCode.workspace = mockWorkspace as unknown as typeof vscode.workspace;
mockVSCode.commands = mockCommands as unknown as typeof vscode.commands;
mockVSCode.window = mockWindow as unknown as typeof vscode.window;

const onNotification = jest.fn();
const catchOnReady = jest.fn();
const catchSendRequest = jest.fn();
const afterOnReady = jest.fn().mockReturnValue({ catch: catchOnReady });
const afterSendRequest = jest.fn().mockReturnValue({ catch: catchSendRequest });
const onReady = jest.fn().mockReturnValue({ then: afterOnReady });
const sendRequest = jest.fn().mockReturnValue({ then: afterSendRequest });
const settingMonitorStart = jest.fn();

const mockExtensionContext = {
	subscriptions: [],
} as unknown as vscode.ExtensionContext;

type ModuleServerOptions = { run: NodeModule; debug: NodeModule };

const isModuleOptions = (options: unknown): options is ModuleServerOptions =>
	Boolean((options as ModuleServerOptions)?.run?.module);

const stripPaths = <T extends unknown[]>(params: T): T => {
	if (Array.isArray(params)) {
		const serverOptions = params[1];

		if (isModuleOptions(serverOptions)) {
			serverOptions.run.module = 'mock-path';
			serverOptions.debug.module = 'mock-path';
		}
	}

	return params;
};

describe('Extension entry point', () => {
	beforeEach(() => {
		jest.clearAllMocks();

		mockVSCode.window.activeTextEditor = undefined;

		mockLanguageClient.mockReturnValue({
			onNotification,
			onReady,
			sendRequest,
		} as unknown as LanguageClient);

		mockSettingMonitor.mockReturnValue({
			start: settingMonitorStart,
		} as unknown as SettingMonitor);

		(mockExtensionContext as { subscriptions: unknown[] }).subscriptions = [];
	});

	it('should provide a public API', () => {
		const api = activate(mockExtensionContext);

		expect(api).toBeInstanceOf(EventEmitter);
	});

	it('should create a language client', () => {
		activate(mockExtensionContext);

		expect(mockLanguageClient).toHaveBeenCalled();
		expect(stripPaths(mockLanguageClient.mock.calls[0])).toMatchSnapshot();
	});

	it('should watch for changes to ec0lint-style configuration files', () => {
		activate(mockExtensionContext);

		expect(mockWorkspace.createFileSystemWatcher).toHaveBeenCalled();
		expect(mockWorkspace.createFileSystemWatcher.mock.calls[0]).toMatchInlineSnapshot(`
		Array [
		  "**/{.ec0lint-stylerc{,.js,.json,.yaml,.yml},ec0lint-style.config.js,.ec0lint-styleignore}",
		]
	`);
	});

	it('should register an auto-fix command', () => {
		const disposable = { dispose: () => undefined };

		mockCommands.registerCommand.mockReturnValueOnce(disposable);

		activate(mockExtensionContext);

		const { subscriptions } = mockExtensionContext;

		expect(mockCommands.registerCommand).toHaveBeenCalled();

		// cspell:enable
		expect(subscriptions).toContain(disposable);
	});

	it('with an active text editor, should send auto-fix commands to the language server', () => {
		window.activeTextEditor = mockTextEditor;

		activate(mockExtensionContext);

		mockCommands.registerCommand.mock.calls[0][1](undefined);

		expect(sendRequest).toHaveBeenCalledTimes(1);
		expect(sendRequest.mock.calls[0]).toMatchSnapshot();
		expect(afterSendRequest).toHaveBeenCalledTimes(1);
		expect(afterSendRequest.mock.calls[0][0]).toBeUndefined();
		expect(mockWindow.showErrorMessage).not.toHaveBeenCalled();
	});

	it('without an active text editor, should not send auto-fix commands to the language server', () => {
		window.activeTextEditor = undefined;

		activate(mockExtensionContext);

		mockCommands.registerCommand.mock.calls[0][1](undefined);

		expect(sendRequest).not.toHaveBeenCalled();
		expect(afterSendRequest).not.toHaveBeenCalled();
		expect(mockWindow.showErrorMessage).not.toHaveBeenCalled();
	});

	it('should show an error message if sending the command request fails', () => {
		window.activeTextEditor = mockTextEditor;

		activate(mockExtensionContext);

		mockCommands.registerCommand.mock.calls[0][1](undefined);
		afterSendRequest.mock.calls[0][1](undefined, new Error());

		expect(sendRequest).toHaveBeenCalledTimes(1);
		expect(afterSendRequest).toHaveBeenCalledTimes(1);
		expect(mockWindow.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(mockWindow.showErrorMessage.mock.calls[0]).toMatchInlineSnapshot(`
		Array [
		  "Failed to apply ec0lint-style fixes to the document. Please consider opening an issue with steps to reproduce.",
		]
	`);
	});

	it('should monitor settings', () => {
		const disposable = { dispose: () => undefined };

		settingMonitorStart.mockReturnValueOnce(disposable);

		activate(mockExtensionContext);

		const { subscriptions } = mockExtensionContext;

		const mockLanguageClientInstance = mockLanguageClient.mock.results[0].value;

		expect(mockSettingMonitor).toHaveBeenCalled();
		expect(mockSettingMonitor.mock.calls[0][0]).toBe(mockLanguageClientInstance);
		expect(mockSettingMonitor.mock.calls[0][1]).toBe('ec0lint-style.enable');
		expect(settingMonitorStart).toHaveBeenCalled();
		expect(subscriptions).toContain(disposable);
	});

	it('should listen for the DidRegisterCodeActionRequestHandler notification', () => {
		activate(mockExtensionContext);

		afterOnReady.mock.calls[0][0]();

		expect(onReady).toHaveBeenCalled();
		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[0][0]).toBe(Notification.DidRegisterCodeActionRequestHandler);
		expect(onNotification.mock.calls[0][1]).toBeInstanceOf(Function);
	});

	it('should set codeActionReady to true when the DidRegisterCodeActionRequestHandler notification is received', async () => {
		const api = activate(mockExtensionContext);

		afterOnReady.mock.calls[0][0]();
		onNotification.mock.calls[0][1]();

		expect(api.codeActionReady).toBe(true);
	});

	it('should listen for the DidRegisterDocumentFormattingEditProvider notification', () => {
		activate(mockExtensionContext);

		afterOnReady.mock.calls[0][0]();

		expect(onReady).toHaveBeenCalled();
		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[1][0]).toBe(
			Notification.DidRegisterDocumentFormattingEditProvider,
		);
		expect(onNotification.mock.calls[1][1]).toBeInstanceOf(Function);
	});

	it('should emit the DidRegisterDocumentFormattingEditProvider event when the DidRegisterDocumentFormattingEditProvider notification is received', async () => {
		const api = activate(mockExtensionContext);

		const promise = new Promise<DidRegisterDocumentFormattingEditProviderNotificationParams>(
			(resolve) => {
				api.on(ApiEvent.DidRegisterDocumentFormattingEditProvider, resolve);
			},
		);

		const params: DidRegisterDocumentFormattingEditProviderNotificationParams = {
			uri: 'file:///foo.css',
			options: {
				documentSelector: [{ scheme: 'file', pattern: '/foo.css' }],
			},
		};

		afterOnReady.mock.calls[0][0]();
		onNotification.mock.calls[1][1](params);

		await expect(promise).resolves.toStrictEqual(params);
	});

	it('should listen for the DidResetConfiguration notification', () => {
		activate(mockExtensionContext);

		afterOnReady.mock.calls[0][0]();

		expect(onReady).toHaveBeenCalled();
		expect(onNotification).toHaveBeenCalled();
		expect(onNotification.mock.calls[2][0]).toBe(Notification.DidResetConfiguration);
		expect(onNotification.mock.calls[2][1]).toBeInstanceOf(Function);
	});

	it('should emit the DidResetConfiguration event when the DidResetConfiguration notification is received', async () => {
		const api = activate(mockExtensionContext);

		const promise = new Promise<void>((resolve) => {
			api.on(ApiEvent.DidResetConfiguration, resolve);
		});

		afterOnReady.mock.calls[0][0]();
		onNotification.mock.calls[2][1]();

		await expect(promise).resolves.toBeUndefined();
	});

	it('should show an error message if the DidRegisterDocumentFormattingEditProvider notification fails', async () => {
		activate(mockExtensionContext);

		await catchOnReady.mock.calls[0][0](new Error('Problem!'));
		await catchOnReady.mock.calls[0][0]('String problem!');

		expect(mockWindow.showErrorMessage).toHaveBeenCalledTimes(2);
		expect(mockWindow.showErrorMessage.mock.calls[0]).toMatchInlineSnapshot(`
		Array [
		  "ec0lint-style: Problem!",
		]
	`);
		expect(mockWindow.showErrorMessage.mock.calls[1]).toMatchInlineSnapshot(`
		Array [
		  "ec0lint-style: String problem!",
		]
	`);
	});
});
