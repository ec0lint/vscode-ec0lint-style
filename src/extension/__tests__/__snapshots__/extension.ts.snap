// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension entry point should create a language client 1`] = `
Array [
  "Stylelint",
  Object {
    "debug": Object {
      "module": "mock-path",
      "options": Object {
        "execArgv": Array [
          "--nolazy",
          "--inspect=6004",
        ],
      },
    },
    "run": Object {
      "module": "mock-path",
    },
  },
  Object {
    "diagnosticCollectionName": "Stylelint",
    "documentSelector": Array [
      Object {
        "scheme": "file",
      },
      Object {
        "scheme": "untitled",
      },
    ],
    "synchronize": Object {
      "fileEvents": undefined,
    },
  },
]
`;
